### PR TITLE
Add optional dependencies to pyproject.toml

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

As suggested in https://github.com/GenericMappingTools/pygmt/pull/1848#discussion_r949080477, this PR adds `optional-dependencies` to pyproject.toml so that people can run `pip install pygmt[all]` to install pygmt with optional dependencies included. 

To test it locally, you can check out this branch, change into the root directory, and run
```
pip install .
```
and 
```
pip install .[all]
```
you should see that `pip install .[all]` also check if geopandas is installed while `pip install .` doesn't.

Address https://github.com/GenericMappingTools/pygmt/pull/1848#discussion_r949080477.

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

**Comment by seisman** [comment]
I'm not sure if we should update the `pip install pygmt` command to `pip install pygmt[all]`
in the installation instructions (https://www.pygmt.org/dev/install.html).

File: pyproject.toml [hunk 1]
```diff
@@ -32,6 +32,11 @@ dependencies = [
 ]
 dynamic = ["version"]
 
+[project.optional-dependencies]
+all = [
+    "geopandas"
```
Comment (weiji14) [inline pyproject.toml:37]: Should we add `ipython` as well as mentioned in https://www.pygmt.org/v0.7.0/install.html#dependencies?
```suggestion
    "geopandas",
    "ipython"
```

**Review by weiji14 — COMMENTED** [review]
> I'm not sure if we should update the `pip install pygmt` command to `pip install pygmt[all]` in the installation instructions (https://www.pygmt.org/dev/install.html).

Wouldn't recommend replacing it with `pip install pygmt[all]`, but I think we can mention the `pip install pygmt[all]` command in https://www.pygmt.org/v0.7.0/install.html#using-pip and indicate that this will install optional dependencies like `geopandas`.

File: pyproject.toml (hunk 1 above)
Comment (seisman) [reply pyproject.toml:37]: Yes, because we also try to import IPython in our codes.

File: doc/install.rst [hunk 2]
```diff
@@ -218,6 +218,11 @@ This installs the latest stable release from
 
     pip install pygmt
 
+.. tips::
```
Comment (seisman) [inline doc/install.rst:221]:
```suggestion
.. tip::
```

File: doc/install.rst [hunk 3]
```diff
@@ -218,6 +218,11 @@ This installs the latest stable release from
 
     pip install pygmt
 
+.. tip::
+
+   You can also run ``pip install pygmt[all]`` to install pygmt alongside with
+   all its optional dependencies.
```
Comment (weiji14) [inline doc/install.rst:224]:
```suggestion
   You can also run ``pip install pygmt[all]`` to install pygmt with
   all of its optional dependencies.
```

**Review by weiji14 — APPROVED** [review]

**Pull request merged**